packages/integration-sdk-runtime/src/synchronization/batchBySize.ts

```diff
@@ -0,0 +1,102 @@
+
+/**
+Converts an array of graphObjects into an array of groups of this objects.
+@param graphObjects  Entities or relationships to group.
+
+@param maximumBatchSizeInBytes The maximum size of the group of objects.
+
+@param logger 
+ */
+export function batchGraphObjectsBySizeInBytes<
+  T extends UploadDataLookup,
+  K extends keyof T,
+>(
+  graphObjects: T[K][],
+  maximumBatchSizeInBytes: number,
+  logger: IntegrationLogger,
+): T[K][][] {
+  const batches: T[K][][] = [];
+
+  let currentBatch: T[K][] = [];
+  let currentBatchSize: number = 0;
+  for (const graphObject of graphObjects) {
+    const typedGraphObject = graphObject as Entity | Relationship;
+
+    let graphObjectSizeInBytes = getSizeOfObject(graphObject);
+
+    if (graphObjectSizeInBytes > maximumBatchSizeInBytes) {
+      const rawDataSizeInBytes = typedGraphObject._rawData
+        ? getSizeOfObject(typedGraphObject._rawData)
+        : 0;
+
+      if (
+        graphObjectSizeInBytes - rawDataSizeInBytes >
+        maximumBatchSizeInBytes
+      ) {
+        logger.warn(
+          {
+            _key: typedGraphObject._key,
+            _type: typedGraphObject.type,
+            maximumBatchSizeInBytes,
+            graphObjectSizeInBytes,
+            rawDataSizeInBytes,
+          },
+          'Graph object is larger than batch size in bytes; cannot upload graph object.',
+        );
+        logger.publishWarnEvent({
+          name: IntegrationWarnEventName.IngestionLimitEncountered,
+          description:
+            'Graph object is larger than batch size in bytes; cannot upload graph object. Will skip graph object.',
+        });
+        continue;
+      } else {
+        logger.info(
+          {
+            _key: typedGraphObject._key,
+            _type: typedGraphObject.type,
+            maximumBatchSizeInBytes,
+            graphObjectSizeInBytes,
+            rawDataSizeInBytes,
+          },
+          'Graph object is larger than batch size in bytes; removing raw data.',
+        );
+        graphObjectSizeInBytes -= rawDataSizeInBytes;
+        typedGraphObject._rawData = undefined;
+      }
+    }
+
+    if (graphObjectSizeInBytes + currentBatchSize >= maximumBatchSizeInBytes) {
+      batches.push(currentBatch);
+      currentBatch = [];
+      currentBatchSize = 0;
+    }
+    currentBatch.push(graphObject);
+    currentBatchSize += graphObjectSizeInBytes;
+  }
+
+  if (currentBatch.length > 0) {
+    batches.push(currentBatch);
+  }
+  return batches;
+}
+
+export function getSizeOfObject(object: any): number {
+  try {
+    return Buffer.byteLength(JSON.stringify(object), 'utf8');
+  } catch (error) {
+    if (error instanceof RangeError) {
+      //If object is too large to size, stringify runs out of memory. We fallback to the max batchSize.
+      //This is highly unlikely since it should only happen after ~500MB.
+      return MAX_BATCH_SIZE_IN_BYTES + 1;
+    }
+    throw error;
+  }
```

Comment:
I"m a little concerned about a small edge case with this logic. Let's say for example, you have a graph object (`graphObject`) for which `Buffer.byteLength(JSON.stringify(object), 'utf8')` returns the following:

- `Buffer.byteLength(JSON.stringify(graphObject), 'utf8')`
   `  => RangeError`
- `Buffer.byteLength(JSON.stringify(graphObject._rawData), 'utf8')`
   `  => 1000`

Because we return `MAX_BATCH_SIZE_IN_BYTES + 1` for all `RangeError`, our logic inside of `batchGraphObjectsBySizeInBytes()` could have us unsafely uploading a resource that is too large.

E.g.

1. `getSizeOfObject(graphObject) `
    `  => MAX_BATCH_SIZE_IN_BYTES + 1`
2. `getSizeOfObject(graphObject._rawData) `
    `  => 1000`
3. `getSizeOfObject(graphObject) - getSizeOfObject(graphObject._rawData)`
     `  => MAX_BATCH_SIZE_IN_BYTES - 999` <= # this is below the `MAX_BATCH_SIZE_IN_BYTES`, and will get uploaded

I would recommend removing the `try/catch`ing inside of this function altogether, and letting the caller`catch` and handle any error from `getSizeOfObject()`.

```suggestion
  return Buffer.byteLength(JSON.stringify(object), 'utf8');
```